modyn/trainer_server/internal/trainer/remote_downsamplers/remote_craig_downsampling.py

```diff
@@ -0,0 +1,196 @@
+    OPTIMIZER_CHOICES,
+)
+
+
+class RemoteCraigDownsamplingStrategy(AbstractPerLabelRemoteDownsamplingStrategy):
+    """
+    Strategy introduced in:
+    Data-efficient Training of Machine Learning Models
+    Implementation adapted from:
+    DEEPCORE https://raw.githubusercontent.com/PatrickZH/DeepCore/main/deepcore/methods/craig.py
+    This strategy selects points via submodular maximization of a per-class FacilityLocation function. The score is
+    proportional to the Euclidean distance between the two samples' gradients.
+    """
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        batch_size: int,
+        params_from_selector: dict,
+        per_sample_loss: Any,
+        device: str,
+    ) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, device)
+
+        self.criterion = per_sample_loss
+
+        self.selection_batch = params_from_selector["selection_batch"]
+        self.greedy = params_from_selector["greedy"]
+        if self.greedy not in OPTIMIZER_CHOICES:
+            raise ValueError(
+                f"The required Greedy optimizer is not available. Pick one of the following: {OPTIMIZER_CHOICES}"
+            )
+
+        # This class uses the embedding recorder
+        self.requires_coreset_supporting_module = True
+
+        # Samples are supplied label by label (this class is instance of AbstractPerLabelRemoteDownsamplingStrategy).
+        # The following list keeps the gradients of the current label. When all the samples belonging to the current
+        # label have been seen, the scores are computed and the list is emptied
+        self.current_class_gradients: list[torch.Tensor] = []
+        # distance_matrix[i,j] = 0 if label[i]!=label[j] else is proportional to the euclidean distance between the
+        # two samples in the gradient space
+        self.distance_matrix = np.zeros([0, 0])
+
+        # if true, the downsampling is balanced across classes ex class sizes = [10, 50, 30] and 50% downsampling
+        # yields the following downsampled class sizes [5, 25, 15] while without balance something like [0, 45, 0] can
+        # happen
+        self.balance = params_from_selector["balance"]
+        if self.balance:
+            self.already_selected_samples: list[int] = []
+            self.already_selected_weights = torch.tensor([]).float()
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        assert embedding is not None
+
+        # Slightly different implementation for BTS and STB since in STB points are supplied class by class while in
+        # BTS are not. STB will always use the first branch, STB will typically (might use the first if all the points
```

Comment:
Not sure what you mean by branch here and also, to my shame, i don't understand the problem here yet. Maybe you can elaborate a bit?

Reply:
Branch refers to the if/else. See the comment above for the explanation.

Reply:
ahh now I understand this comment. maybe you can extend this a bit with the comment you wrote above. I think there is also a typo here because you mention STB twice in the second sentence

Reply:
Yes, you're right. Fixed